### ISSUE TITLE
Encrypting/decrypting very slow due to use of de-/encryptWithPassword()

### ISSUE THREAD

**Issue by SunMar** [thephpleague/oauth2-server#812]
Hi,

During tests we discovered that encrypting and decrypting is very slow. This is caused by the use of the [PBKDF2 algorithm](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/KeyOrPassword.php#L83) in [`KeyOrPassword`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/KeyOrPassword.php#L7) (from the [`defuse/php-encryption`](https://github.com/defuse/php-encryption) dependency), which can be traced back to [`CryptTrait`](https://github.com/thephpleague/oauth2-server/blob/925776958fc3f5278e74363663c20147af32b668/src/CryptTrait.php#L14) where the [`encryptWithPassword()`](https://github.com/thephpleague/oauth2-server/blob/925776958fc3f5278e74363663c20147af32b668/src/CryptTrait.php#L32) and [`decryptWithPassword()`](https://github.com/thephpleague/oauth2-server/blob/925776958fc3f5278e74363663c20147af32b668/src/CryptTrait.php#L49) methods are used. The use of this algorithm is secure if you're unsure about how good your encryption key is, however if you know you have a secure encryption key, then PBKDF2 is not needed.

An easy way to get a big performance boost here is by allowing a [`Key`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Key.php#L7) object to be used as encryption key instead of a string. If it's a [`Key`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Key.php#L7) object then [`encrypt()`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Crypto.php#L20) instead of [`encryptWithPassword()`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Crypto.php#L41) and [`decrypt()`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Crypto.php#L62) instead of [`decryptWithPassword()`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Crypto.php#L84) can be used. With a [`Key`](https://github.com/defuse/php-encryption/blob/5176f5abb38d3ea8a6e3ac6cd3bbb54d8185a689/src/Key.php#L7) the encryption library trusts the key and skips PBKDF2 using only the much faster [HKDF algorithm](https://github.com/defuse/php-encryption/blob/893217037f3ec62c542738189a8b90169275034a/src/KeyOrPassword.php#L64).

This could be implemented backwards compatible by checking in `CryptTrait` if `$this->encryptionKey` is an instance of `Key` or not. I can see to make a pull request for this but would first like to know if something like this would be accepted.

**Comment by Sephster** [comment]
In principle I can't think of an issue with what you are proposing as long as it is backwards compatible but it would be good to get more information. If you can provide any links that can back up your assertions that would be appreciated. I noted from the php-encryption lib that they note PBKDF2 is slower but I didn't find much in the way of pros and cons between password vs key other than that.

Is the use of `encryptWithPassword` adversely affecting your production system? If you could provide some more context about the issue being faced and any benchmarks you have got from your tests as well that would really help us to make a decision. Thanks for your time

**Comment by SunMar** [comment]
See defuse/php-encryption#359 for information about PBKDF2 being slow vs using a `Key`. Unfortunately I can't give any benchmarks but our performance tests showed that the PBKDF2 hashing algorithm was forming a big bottleneck, taking up to 50% of a request's total processing time. Once we put some hacks in place to use a `Key` instead, the bottleneck disappeared. The HKDF hashing did not take significantly more time than any other code being run during the request.

**Comment by SunMar** [comment]
Also see defuse/php-encryption#161 which is the original use case for implementing the PBKDF2 algorithm. From what I understand it's for cases where you don't store the key on the server, but have it supplied by the user in the form of a password. PBKDF2 is then used to turn the password into an encryption key and it's meant to be slow to protect against attacks like brute force. However if you are storing the encryption key on the server, and it's a secure encryption key then PBKDF2 is not necessary, only HKDF should be secure enough. The `Key` object by default generates keys that are 32 bytes long and uses bytes, so it's not limited to human readable characters like a password is. The key passed to the `AuthorizationServer` should be such a `Key` I think anyways and not a password.

**Comment by SunMar** [comment]
Perhaps when this is merged and in a release, the following documentation should be updated too to reflect the ability to use a `Key`:

[V5 Security Improvements#6.0.0](https://oauth2.thephpleague.com/v5-security-improvements/)

[Installation#Generating encryption keys](https://oauth2.thephpleague.com/installation/)

From skimming through he [documentation](https://oauth2.thephpleague.com/) I couldn't find other places referencing encryption keys.

**Comment by SunMar** [comment]
Ok, I found the `gh-pages` branch and also created a PR #820 for that branch to reflect the changes in the documentation, if they are accepted.

**Comment by sg3s** [comment]
Hi,

I just wanted to note that I encountered this exact issue too, when I was implementing the library directly in my own project for different purposes (after having found it this project).

In my case when decrypting just one item while using `decryptWithPassword` the PBKDF2 overhead accounted for almost 200ms of the process time for a request that took ~260ms total, which was ~60ms without decrypting the value (profiled with xdebug / qcachegrind).

After switching to the methods described in the [defuse/php-encryption documentation](https://github.com/defuse/php-encryption/blob/master/docs/Tutorial.md#scenario-1-keep-data-secret-from-the-database-administrator) this slowdown essentially disappeared.

Checkout the neat bin script the lib provides to generate a key, might want to include that in the docs maybe?

The performance hit is significant, I came here to see if it was a known problem. Good to see someone already beat me to it. Would be good to see this updated upstream :)
